pmlc/transforms/make_entrypoints.cc

```diff
@@ -0,0 +1,160 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+
+#include "pmlc/transforms/pass_detail.h"
+#include "pmlc/util/ids.h"
+
+namespace LLVM = mlir::LLVM;
+using LLVMType = LLVM::LLVMType;
+
+namespace pmlc::transforms {
+namespace {
+
+class MakeEntrypointsPass final
+    : public MakeEntrypointsPassBase<MakeEntrypointsPass> {
+public:
+  void runOnOperation() final;
+
+private:
+  mlir::Location getLoc() { return getOperation().getLoc(); }
+  void initMembers();
+  mlir::FuncOp getNetworkEntry();
+
+  void makePlaidmlInit(mlir::OpBuilder &builder);
+  void makePlaidmlExecute(mlir::OpBuilder &builder, mlir::FuncOp entryFunc);
+  void makePlaidmlFini(mlir::OpBuilder &builder);
+
+  LLVMType llvmInt32Type;
+  LLVMType llvmPtrTy;
+  LLVMType llvmVoidTy;
+};
+
+void MakeEntrypointsPass::runOnOperation() {
+  mlir::FuncOp entryFunc = getNetworkEntry();
+  if (!entryFunc) {
+    return;
+  }
+
+  initMembers();
+  mlir::OpBuilder builder(getOperation().getBody()->getTerminator());
+  makePlaidmlInit(builder);
+  makePlaidmlExecute(builder, entryFunc);
+  makePlaidmlFini(builder);
+}
+
+void MakeEntrypointsPass::initMembers() {
+  auto ctx = &getContext();
+  llvmInt32Type = LLVMType::getInt32Ty(ctx);
+  llvmPtrTy = LLVMType::getInt8PtrTy(ctx);
+  llvmVoidTy = LLVMType::getVoidTy(ctx);
+}
+
+mlir::FuncOp MakeEntrypointsPass::getNetworkEntry() {
+  mlir::FuncOp entryFunc;
+
+  getOperation().walk([&](mlir::FuncOp func) {
+    if (func.isExternal() || func.getName() != networkMain) {
+      return mlir::WalkResult::advance();
+    }
+    if (entryFunc) {
+      getOperation().emitError(
+          "Expected only one entrypoint function definition");
+      signalPassFailure();
+      return mlir::WalkResult::interrupt();
+    }
+    entryFunc = func;
+    return mlir::WalkResult::advance();
+  });
+  if (!entryFunc) {
+    getOperation().emitError(
+        "Expected a single entrypoint function definition");
+    signalPassFailure();
+  }
+
+  return entryFunc;
+}
+
+void MakeEntrypointsPass::makePlaidmlInit(mlir::OpBuilder &builder) {
+  // This method builds:
+  //
+  //   Network* plaidml_init(Device* device) {
+  //     return device;  // Network == Device for now.
+  //   }
+  //
+  // This is a rather trivial function; the plan is to expand it later.
+  // In particular, the Network parameter gives networks a mechanism to carry
+  // additional information from plaidml_init to plaidml_exec and
+  // plaidml_fini.
+  auto func = builder.create<mlir::FuncOp>(
```

Comment:
I'd expect to be operating in the LLVM dialect and to create `llvm::FuncOp` instead of `mlir::FuncOp`

Reply:
Alternatively, I could see this split happening at PXA dialect or comp dialect level, and then having a straightforward legalization to convert high-level `mlir::FuncOp` into low level `llvm::FuncOp` (converting a `Network` data type in to a `void*` for instance).